pkg/v13/kubeconfig/desired.go

```diff
@@ -0,0 +1,57 @@
+package kubeconfig
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/kubeconfig"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/client/k8srestconfig"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (r *StateGetter) GetDesiredState(ctx context.Context, obj interface{}) ([]*corev1.Secret, error) {
+	cr, err := ToCustomObject(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	appOperator, err := r.certsSearcher.SearchAppOperator(cr.ClusterID)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	c := k8srestconfig.Config{
+		Logger: r.logger,
+
+		Address:   cr.APIDomain,
+		InCluster: false,
+		TLS: k8srestconfig.ConfigTLS{
+			CAData:  appOperator.APIServer.CA,
+			CrtData: appOperator.APIServer.Crt,
+			KeyData: appOperator.APIServer.Key,
+		},
+	}
+	restConfig, err := k8srestconfig.New(c)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	yamlBytes, err := kubeconfig.NewKubeConfigForRESTConfig(ctx, restConfig, fmt.Sprintf("giantswarm-%s", cr.ClusterID), "")
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	secret := corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.resourceName,
+			Namespace: r.resourceNamespace,
+		},
+		Data: map[string][]byte{
+			"kubeConfig": yamlBytes,
+		},
+	}
+
+	return []*corev1.Secret{&secret}, nil
```

Comment:
What if the `cr` deletion timestamp is set? Shouldn't this secret be deleted then? How to ensure that the secret stays as long as it's needed?

Reply:
@kopiczko  Do we need to care that information at `desired.go`?

Reply:
Yes, how will you delete the secret otherwise?

Reply:
I tried to have the same practice as in here. https://github.com/giantswarm/app-operator/blob/ef98e4bd1a973680ba1ac553119dcb4f823ad666/service/controller/appcatalog/v1/resource/index/desired.go
The operatorkit/secret package would do those remaining things incl. deleting CRs.

Reply:
Yes you did it there: https://github.com/giantswarm/app-operator/blob/ef98e4bd1a973680ba1ac553119dcb4f823ad666/service/controller/appcatalog/v1/resource/index/desired.go#L44-L46. Not I don't se it here.

Reply:
🤦‍♂️ My bad

Reply:
Can we check that after this PR, I would like to put that code inside of each provider pkg. (e.g. aws, azure)

Reply:
I think I don't follow. Why not just adding the `if`?

Reply:
For that matter, I need to check `ObjectMeta` but since this method had been changed to get `ClusterConfig` as a parameter, and this is not a CR. 
I can get that information when I transformed it from actual `pkg`, will show you after merge this :)

Reply:
You can do something like that: https://github.com/giantswarm/release-operator/pull/76/files#diff-df2daa4144c3c1cc61822e503606f4d6R24

Reply:
Thanks, `ClusterConfig` is a custom struct in this file, so I can't get the `deletionTimestamp`, which reside inside CR instance. Can I do that in `/pkg`? 

if would look like below, 
```golang
func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error) {
  // Turn object into CR
  cr := ToCustomObject(obj)
  if IsDeleted(cr) {
    return nil, nil
  }
  config := aws.GetClusterConfig(cr)
  result, err := kubeconfig.GetDesiredState(context, config)
  if err != nil {
  ..
  }
  return result, nil
}
```

Reply:
I think this is confusing. I think this object should take `GetClusterConfigFunc` as a parameter in the `Config` and call it on its own.